File: algorithms/19.swift
// https://leetcode.com/problems/remove-nth-node-from-end-of-list

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func removeNthFromEnd(head: ListNode?, _ n: Int) -> ListNode? {
        var n = n
        var me = head
        var pioneer = head

        // setup pioneer
        while n > 0 && pioneer != nil {
            pioneer = pioneer!.next
            n -= 1
        }
        if pioneer == nil {
            if n == 0 { return head!.next }
            else { return head }
        }

        // march
        while pioneer!.next != nil {
            pioneer = pioneer!.next
            me = me!.next
        }

        me!.next = me!.next!.next
        return head
    }
}

File: algorithms/19.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """

        me = head
        pioneer = head

        # first, place pioneer n-steps forward
        while n > 0 and pioneer is not None:
            pioneer = pioneer.next
            n -= 1
        if pioneer is None:
            return head.next if n == 0 else head

        # march
        while pioneer.next is not None:
            pioneer = pioneer.next
            me = me.next
        me.next = me.next.next

        return head
